### DISCUSSION TITLE
Using the container registry as a proxy prevents pushing images to it?

### DISCUSSION THREAD

**Discussion by flixman** [distribution/distribution#4460] [Q&A]
I have a private registry to which I can push/pull containers successfully. As I want to reduce the queries/traffic to docker hub while doing development, yesterday I enabled the proxy feature through the container environment variable ```REGISTRY_PROXY_REMOTEURL=https://registry-1.docker.io```. This works as expected for the pulls (both the images available locally on the registry and in docker.io can be pulled, with the latter ones being cached by the registry), but I have lost all possibility to push to my registry (I am getting a 500 Internal Server Error, and on the registry logs I am getting a 401 message with the description "denied: requested access to the resource is denied").

Is this expected behavior?

**Comment by amir-bialek** [comment]
Hey @flixman 
When using the registry as proxy, it is not possible to push images to it.
This is normal behavior..

We have the same status as you, our CI is pushing the images always to dockerhub, and the local env is pulling it from our registry which is proxy-mirror to dockerhub.  it is working well.

**Reply by flixman** [reply]
@amir-bialek But this is a bit weird: Should I switch from proxy mode to normal mode, than I can push the images to it. After that I switch it back to mirror and I can pull those images, meaning: considering the fact that I am getting a 500 error from the mirror, this seems more a use case that was not considered (and that guards were not put in to prevented) rather than intended behavior.

What would be the "solution" then? To have a private registry and the mirror, thus duplicating the space for the repo? That is not nice, neither :-/

**Reply by milosgajdos** [reply]
> But this is a bit weird

It's not weird. No. You don't push things into *caching* proxies: the cache is populated by pulling things out of the store. This is pretty much the standard behaviour of any caching proxies, not just registry cache proxy.

**Reply by flixman** [reply]
but this is not entirely a "caching proxy", right? On a caching proxy I'd be able to push and pull, meaning that the proxy becomes a local mirror of the remote store. In this case, this is is advertised as a "pull through cache"... without specifying what will happen to the pushes. If I am running a local register, to which I push and pull images, and I read about the "pull through cache" feature, anybody will read it as a feature extending the existent configuration (as opposed to replacing it, which is what actually happens).

**Reply by amir-bialek** [reply]
I am not sure about the exact definition, but for us it is working well. 
We are pushing to a private repo in docker-hub, and pulling from our private registry. 
I believe it is important to have the images on docker-hub for back, incase something will happen to our cluster.
@milosgajdos  btw we upgraded to v3-beta, so far so good.

**Reply by milosgajdos** [reply]
> but this is not entirely a "caching proxy", right? On a caching proxy I'd be able to push and pull

Honestly, I don't know what you're talking about here. It literally is a caching proxy: it caches container images.

Take Cloudflare caching for example: you smack a GET to some content and CF can cache it, but it will never cache POST/PUT data. 

I honestly don't know of any caching proxy to which you can "push data".
